### DISCUSSION TITLE
Sidecar pattern for SQL DB connection with Keycloak CR

### DISCUSSION THREAD

**Discussion by danielgit0** [keycloak/keycloak#25749] [Archive]
Hi, the plan is to do a deployment in GKE and [connect to a Cloud SQL instance](https://cloud.google.com/sql/docs/mysql/connect-kubernetes-engine#workload-identity). They provide many [reasons to use a sidecar](https://cloud.google.com/sql/docs/mysql/connect-kubernetes-engine#run_the_in_a_sidecar_pattern).

Is there a way to achieve the sidecar pattern with the Keycloak CR? Would it work with the [Pod Template](https://www.keycloak.org/operator/advanced-configuration)?
So far I couldn't find any related info/docs/discussions, any hint would be helpful.

Thanks in advance.

**Comment by danielgit0** [comment]
Hi @vmuzikar  / @shawkins / @andreaTP, for sure you already know this. I had a look at the `podTemplate`, but there are some validations that won't allow to do add a side container. It is just to extend and merge some additional configurations to the Keycloak pod that will be created.

Would it be possible to extend it at some point to do it like [Kong - Deploying Sidecars](https://docs.konghq.com/gateway-operator/1.1.x/customization/sidecars/)? The warning "Use at your own risk..." doesn't give me any hope.

If not, then I guess that I could give it a try to [istio's sidecar injection](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/) and [control the injection policy at pod level](https://istio.io/latest/docs/setup/additional-setup/sidecar-injection/) by adding the required label with the `podTemplate` to don't mess up with the Keycloak Operator and dealing with some [istio cloud sql configuration](https://github.com/istio/istio/issues/6593#issuecomment-420591213). 
<img width="712" alt="Bildschirmfoto 2023-12-22 um 10 22 37" src="https://github.com/keycloak/keycloak/assets/51970426/df4d3c9f-4e72-4109-a1d8-ede3a2284022">

What do you think?

**Reply by vmuzikar** [reply]
> but there are some validations that won't allow to do add a side container

@danielgit0 Could you please elaborate what validations you mean specifically? The limitation that Keycloak is always the first container?

**Reply by danielgit0** [reply]
hi @vmuzikar

First, thank you for taking the time to have a look and answer.

It was my bad, now that you point "... Keycloak is always the first container...", I had a look again and realized that I completely overlooked the `.findFirst()` in the PodTemplate validation.

Here is a snippet of a small test where it shows that it is possible to add another container without any problems as long as it is not the first one:

```
  unsupported:
    podTemplate:
      spec:
        containers:
          - { }
          - name: hello-node
            image: registry.k8s.io/e2e-test-images/agnhost:2.39
```

And the pod with the 2 containers running:
![Bildschirmfoto 2024-01-02 um 13 36 33](https://github.com/keycloak/keycloak/assets/51970426/ddf24c03-cc7c-44c4-a9c5-e48e2662cf5f)


I was always trying to add the sidecar using the first container. At some point, it wasn't necessary to add any additional configuration for the Keycloak container, but and empty object as the first element does the trick too.

**Reply by shawkins** [reply]
> empty object as the first element does the trick too

We may eventually want to revisit that and use the container name, rather than the position, to match / merge. That will make the application of the unsupported spec more similar to a strategic merge patch.

**Reply by vmuzikar** [reply]
@shawkins +1
We can certainly consider that as a future enhancement.
